### PR TITLE
Add solution to Assignment 5

### PR DESCRIPTION
**Pull request by elifguler96**

### CODE REVIEW

File: Assignment5.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                    el.decrementindegree();
+                }
+            }
+        }
+
+        return set;
+    }
+}
+
+class Vertex {
```
Comment (adifeiner) [inline Assignment5.java]: As I learned from @brzozan, class definitions should be in the beginning.

File: Assignment5.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+     *
+     * @param words list of words in lexicographic order.
+     * @return characters in lexicographic order.
+     */
+    public static List<Character> inferAlphabet(List<String> words) {
+        if (words.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        Set<Character> set = new HashSet<>();
```
Comment (adifeiner) [inline Assignment5.java]: You can also add all the characters to the graph instead of to the set and then add the edge between the vertices. If you don't use set variable you don't need to check lines 101 - 107 in "createEdgeBetweenVertices" and line 49 in "inferAlphabet".

File: Assignment5.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+     * @return topologically sorted set of characters.
+     */
+    private static Set<Character> topologicalSort(Map<Character, Vertex> graph) {
+        Set<Character> set = new LinkedHashSet<>();
+
+        if (graph.isEmpty()) {
+            return set;
+        }
+
+        Vertex[] vertices = new Vertex[graph.size()];
```
Comment (adifeiner) [inline Assignment5.java]: It seems that this array is not really necessary since you can loop over the graph instead of the array in the while loop. Is it influencing the efficiency of the program? (in terms of space efficiency)
@brzozan

**Review by adifeiner — COMMENTED** [review]
I really like your approach for finding the difference between the words!

File: Assignment5Test.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    void inferAlphabetTestNormal() {
+        List<String> words = Arrays.asList("ART", "RAT", "CAT", "CAR");
+        List<Character> alphabet1 = Arrays.asList('A', 'T', 'R', 'C');
+        List<Character> alphabet2 = Arrays.asList('T', 'A', 'R', 'C');
+        assertTrue(inferAlphabet(words).equals(alphabet1) || inferAlphabet(words).equals(alphabet2));
+    }
+
+    @Test
+    void inferAlphabetTestOneWord() {
```
Comment (brzozan) [inline Assignment5Test.java:22]: This is an interesting test for an edge case!

You could simplify it a bit by either converting results to a set and comparing it with another set that contains expected elements.

Another possibility: check the size of result and manually check if it contains all expected letters using `contains(Object o)`.

File: Assignment5.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        Set<Character> set = new HashSet<>();
+
+        String s1 = words.get(0);
+        // Performed so that all characters will be in the returned list even if topologicalSort doesn't contain some.
+        addAllCharacters(s1, set);
+
+        Map<Character, Vertex> graph = new HashMap<>();
+
+        for (int i = 0; i < words.size() - 1; i++) {
+            s1 = words.get(i);
```
Comment (brzozan) [inline Assignment5.java]: Be careful with using `get(index)` on a "plain" `List` - someone could pass `LinkedList` as parameter, which would give you O(n^2) complexity to just traverse the list.

The best solution would be to use `Iterator` (or `ListIterator`, if its features would be useful). Quite a bit worse solution would be to require `ArrayList` as parameter to have guarantees about complexity of `get(index)`.

File: Assignment5.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    private static int findFirstDifferingIndex(String s1, String s2) throws NoDifferenceException {
+        int bound = Math.min(s1.length(), s2.length());
+        int index = 0;
+
+        while (index < bound && s1.charAt(index) == s2.charAt(index)) {
+            index++;
+        }
+
+        if (index == bound) {
+            throw new NoDifferenceException();
```
Comment (brzozan) [inline Assignment5.java]: I feel like using exception does not really fit here - you define special class to signal it, but then ignore it if exception is thrown. Exceptions should be used to handle things that are, well, exceptional (like errors), while it looks like you couldn't decide here whether identical strings are a problem (throw exception) or not (ignore it).

If two identical strings in dictionary are OK, I think it would be better to either use for example `Optional<Integer>` as return value or use `-1` as special value indicating "no differences".

If two identical strings are not OK, just throwing `IllegalArgumentException` and letting program die would be acceptable.

File: Assignment5.java (hunk 1 above)
Comment (brzozan) [reply Assignment5.java]: Or, to be more precise: this particular one should either be in a separate file or should be inner class of `Assignment5`.

**Review by brzozan — CHANGES_REQUESTED** [review]
Sorry for late review.

It is a really nice, clean and well tested PR. I added a few comments in addition to the ones by Adi.

File: Assignment5.java (hunk 2 above)
Comment (elifguler96) [reply Assignment5.java]: Thanks for the heads up :)

File: Assignment5.java (hunk 6 above)
Comment (elifguler96) [reply Assignment5.java]: Should I use an exception only if there is a problem?

File: Assignment5.java (hunk 3 above)
Comment (elifguler96) [reply Assignment5.java]: Thanks :)

File: Assignment5Test.java [hunk 7]
```diff
@@ -19,15 +21,8 @@ void inferAlphabetTestNormal() {
...
-        List<Character> alphabet2 = Arrays.asList('T', 'A', 'R');
-        List<Character> alphabet3 = Arrays.asList('A', 'R', 'T');
-        List<Character> alphabet4 = Arrays.asList('R', 'A', 'T');
-        List<Character> alphabet5 = Arrays.asList('R', 'T', 'A');
-        List<Character> alphabet6 = Arrays.asList('T', 'R', 'A');
-        assertTrue(inferAlphabet(words).equals(alphabet1) || inferAlphabet(words).equals(alphabet2) ||
-                inferAlphabet(words).equals(alphabet3) || inferAlphabet(words).equals(alphabet4) ||
-                inferAlphabet(words).equals(alphabet5) || inferAlphabet(words).equals(alphabet6));
+        Set<Character> alphabet = new HashSet<>(Arrays.asList('A', 'T', 'R'));
+        assertTrue(alphabet.containsAll(inferAlphabet(words)));
```
Comment (brzozan) [inline Assignment5Test.java]: It would be good to also check size, since technically this test would pass if `inferAlphabet(words)` returned  only 'A'.

**Review by brzozan — APPROVED** [review]
LGTM. Thank you!
